[website/docs]: Add Functions section with info on ctx/self (#937)
ethereum/fe: docs/src/spec/functions/context.md

# Context

> `Context` is used frequently in Fe smart contracts. It is used to gate access to EVM functions for reading and modifyng the blockchain.

Author: ```suggestion
`Context` is used frequently in Fe smart contracts. It is used to gate access to EVM features for reading and modifyng the blockchain.
```

## Rationale

Smart contracts execute on the Ethereum Virtual Machine (EVM). The EVM exposes features that allow smart contracts to query or change some of the blockchain data, for example emitting logs that are included in transaction receipts, creating contracts, obtaining the current block number and altering the data stored at certain addresses.

To make Fe maximally explicit and as easy as possible to audit, these functions are gated behind a `Context` object. This is passed as an argument to functions, making it clear whether a function interacts with EVM features from the function signature alone. 

For example, the following function looks pure from its signature (i.e. it is not expected to alter any blockchain data) but in reality it does modify the blockchain (by emitting a log).

```rust
struct SomeEvent{
}

> pub fn looks_pure_but_isnt() {

Author: ```suggestion
pub fn looks_pure_but_isnt() {
    // COMPILE ERROR
```

  emit(SomeEvent())
}
```

Using `Context` to control access to EVM functions such as `emit` solves this problem by requiring an instance of `Context` to be passed explicitly to the function, making it clear from the function signature that the function executes some blockchain interaction. The function above, rewritten using `Context`, looks as follows:

```fe

> struct SomeEvent {

Author: ```suggestion
```

> }

Author: ```suggestion
```

> 

Author: ```suggestion
```

> pub fn uses_context(mut ctx: Context) {

Author: ```suggestion
pub fn uses_context(ctx: Context) {
```

>     ctx.emit(SomeEvent())

Author: ```suggestion
    return ctx.block_number()
```

}
```

## The Context object

The Context object gates access to features such as:

- emitting logs
- creating contracts
- transferring ether
- reading message info
- reading block info

The `Context` object needs to be passed as a parameter to the function. The `Context` object has a defined location in the parameter list. It is the *first* parameter unless the function also takes `self`. `Context` or `self` appearing at any other position in the parameter list causes a compile time error.

The Context object is automatically injected when a function is called externally but it has to be passed explicitly when the function is called from another Fe function e.g.

```fe
// The context object is automatically injected when this is called externally
pub fn multiply_block_number(ctx: Context) -> u256 {
  // but it has to be passed along in this function call
  return retrieves_blocknumber(ctx) * 1000
}

fn retrieves_blocknumber(ctx: Context) -> u256 {
  return ctx.block_number()
}
```

## Context mutability

> All functionality that modifies the blockchain such as creating logs or contracts or transferring ether would require a mutable `Context` reference whereas read-only access such as `ctx.blocknumber()` does not need require a mutable reference to the context. To pass a mutable `Context` object, prepend the object name with `mut` in the function definition, e.g.:

Author: ```suggestion
All functionality that modifies the blockchain such as creating logs or contracts or transferring ether would require a mutable `Context` reference whereas read-only access such as `ctx.block_number()` does not need require a mutable reference to the context. To pass a mutable `Context` object, prepend the object name with `mut` in the function definition, e.g.:
```
